### ISSUE TITLE
investigate Language Server Protocol for Merlin

### ISSUE THREAD

**Issue by avsm** [ocaml/merlin#607]
https://github.com/Microsoft/language-server-protocol/wiki/Protocol-Implementations

There seems to be a sophisticated [OCaml implementation](https://github.com/freebroccolo/ocaml-language-server) here as well, that already interops with Merlin.  Should this be mentioned and/or tested more with Merlin?

**Comment by let-def** [comment]
It is worth mentioning. I will ask @freeboccolo what he thinks of merging some features inside Merlin.

**Comment by freebroccolo** [comment]
@let-def I would be happy to see language server support become part of merlin somehow.

I think there are two aspects to consider here:

1. merlin speaking the language server protocol
2. merlin supporting the features specified by the language server capabilities

For (1), this might be the right direction to move merlin in the long term. I'm not sure if it's a good idea for now though, at least not as the primary interface. The problem is that while there are several plugins for language server protocol support being developed for various editors, most of them except for vscode are not yet in a usable state.

But for the long term, it may make sense. Using JSON-RPC for communication with merlin seems fine to me and there's no problem adding custom notifications or requests specific to merlin on top of what the language server protocol already specifies.

For (2), there are some features I've built on top of merlin in the ocaml language server in order to support the full server capabilities specified by the protocol. I think the main ones are:

1. code action provider
2. code lens provider
3. workspace symbol provider

The code action feature (1) just analyzes certain error messages from OCaml (via merlin) and suggests various fixes. Some of these fixes can be applied automatically. The typical case would be like a simple syntax error due to a missing delimiter or something like that. But they don't have to be errors either. They could just be suggestions to fix certain kinds of warnings too. Having merlin be able to support a feature like this directly would be nice.

The code lens feature (2) allows the editor to display information above some text range. This can be used for all sorts of things but what I have done with it in the language server is display types:

![code-lens](https://cloud.githubusercontent.com/assets/151197/20631485/ff4bf51c-b2f3-11e6-91b8-b77901c2f3a3.png)

The feature is implemented by making an outline request to merlin, using the response to find the ranges of let bindings, and then asking merlin for the types. This could be simplified if merlin could also provide the types of structure items in the outline response or if it supported some other special code lens command.

The workspace symbol provider feature (3) allows the user to search for symbols across all the files in the current workspace. I implement this by asking merlin about the source directories it knows about (aside from `~/.opam`), getting the outline data for all of the modules in those directories, and storing the ranges in a database. When files are modified, symbols for that file are refreshed in the database. In other words, (3) is a feature kind of similar to what `ocp-browser` provides.

One other feature that language servers can support that I have not yet implemented is a workspace-wide "find references". This will require getting the symbols for all of the documents, looking up where they are defined, and storing it in the index like with (3).

So if merlin could provide some sort of builtin support for these things (maybe with help from ocp-index for some parts), that would be worth exploring.

Now, the current `ocaml-language-server` is implemented in typescript targeting node. I have done it this way because MS provides libraries for the initialization and messaging functionality for the client and server and this was the quickest way to get a solid implementation up and running.

I have done some work on porting the server to pure OCaml but it's not ready just yet and there are a few other issues I wanted to fix up first.

So what I might suggest as a way to move forward here is to take a look at whether the features I described could be added to merlin. In the meantime, I will try to finish up some parts of the current implementation and work on getting the OCaml server in a usable state. Once that is further along, then it might be a good time to consider whether we can merge the OCaml server directly into merlin or something along those lines.

Thoughts?

**Comment by let-def** [comment]
1) Ok for the language server support, I also had the impression that support was quite limited too.
The next version of Merlin will introduce a "stateless protocol": the `ocamlmerlin` binary can be invoked like a plain unix command, reading all information from command line and stdin, answering on stdout, logging on stderr. Merlin will fork and manage cache in the background.
Of course the JSON-RPC interface is still available. And LSP can come later :).
 
2) Code action feature is interesting. A first step would be to list actions that are worth suggesting (and maybe abstract the langage used to support OCaml and Reason), then discuss how to implement that.

Code lens: have you considered using type enclosing request to implement it? A single request should provide enough information, and merlin could report more information if necessary.

**Comment by bluddy** [comment]
Glad to hear about the stateless protocol being in the next version, as that will also help add support to neovim's neomake plugin.

**Comment by freebroccolo** [comment]
> Ok for the language server support, I also had the impression that support was quite limited too.
> The next version of Merlin will introduce a "stateless protocol": the ocamlmerlin binary can be invoked like a plain unix command, reading all information from command line and stdin, answering on stdout, logging on stderr. Merlin will fork and manage cache in the background.
> Of course the JSON-RPC interface is still available. And LSP can come later :).

Sounds good. For the current language server implementation, I think it probably makes sense to continue using the JSON interface since it's not difficult for us to manage the process with node and I imagine the latency may be a little better for some chains of rapid requests we make for some complex operations like symbol indexing or the code lens.

> Code action feature is interesting. A first step would be to list actions that are worth suggesting (and maybe abstract the langage used to support OCaml and Reason), then discuss how to implement that.

Also sounds good. I can put together a list of the current messages we handle (there are only a few) and maybe open a separate issue to discuss the general approach.

> Code lens: have you considered using type enclosing request to implement it? A single request should provide enough information, and merlin could report more information if necessary.

The way the code lens feature works in the language server protocol is broken up into two parts:

First, the editor will make a request to the server at some point for a list of code lens items for the entire document. You can at that point choose to provide the full information for each code-lens entry, or you can instead provide it with a "thunk", which is only resolved once the code-lens decoration actually comes into view in the editor buffer.

So what I do is gather up all of the top-level module items with outline, and then only when a particular code-lens is resolved (the "thunk") do I make the type enclosing request.

The current approach works quite well in practice, but it's maybe not the most efficient way to do it since I have to make several requests to merlin, i.e., asking for the outline for each module, then maybe several type enclosing requests depending on what is currently visible in the buffer. 

However, merlin has already computed the types by the time the first outline request is called, so the only thing we gain by using this 2-step approach in practice is we potentially avoid some unnecessary type enclosing requests.

If we could just ask for the types of certain kind of module items (along with decoration ranges) in the first place (instead of using outline + type enclosing), it would be more direct.

But this is just one possible use for the code lens feature with merlin so it may be worth more investigation first like the code actions feature.

> Symbol provider: I did a proof of concept of indexing from merlin side. It should be possible to provide a more convenient command for that feature. Would listing the content of all cmis in build directories be a good feature? Maybe excluding the one from packages.

Yes, that sounds like it would be very useful. What I am currently doing is just literally making an outline request for each module, and then I have to manage the database myself. So what you suggest would definitely be an improvement.

> Find references is harder to implement. I had trouble finding a good definition of "this is the same reference" (locate is full of tricks). We can try to implement a naive version first and refine the semantics progressively.

Okay. For comparison, most of the language servers that implement workspace wide find references seem to have some delay in computing the response anyway, so I think it's fine even if we can't make it too efficient at first.

**Comment by freebroccolo** [comment]
One more thing I should mention since it could have an impact on merlin design decisions: the language server protocol is intended to facilitate asynchronous communication between the editor and server. Because of this, requests must have an `id` field and responses must contain an `id` field matching the value of the corresponding request.

The main reason for doing it this way is because for the most full-featured servers, there may be quite a lot of requests made in rapid succession and some of them may take some time to finish (e.g., find references). But in the meantime, other requests should be able to complete as soon as possible.

What I have done in the current language server implementation is to provide an asynchronous priority queue interface on top of merlin. It's not the same as having merlin operate this way internally, but it does allow us to do things like queue a bunch of requests for non-critical data, like outline for code lens, but then still have type-on-hover and completion requests be processed almost as soon as possible by assigning them higher priority.

**Comment by freebroccolo** [comment]
@let-def There is another feature that I forgot to mention called "signature help". Basically what it does is it shows a hover widget indicating the current argument during function application. For languages supporting overloaded functions/methods, it also has a widget to toggle between the different possible signatures for the current application.

I have prototyped this feature for Reason and I think it would work nicely, but the current implementation is a little bit tricky. I wanted to use the `enclosing` command for this (to find the head of the application) but it turns out that the positions it returns are incorrect in many cases, at least for Reason (/cc @jordwalke).

Anyway, it would probably be easier to support as a builtin command.

Below is a demo of what it looks. The underlined type indicates the current argument.

![signature-help](https://cloud.githubusercontent.com/assets/151197/21400580/0b577c12-c76e-11e6-8da6-0d1b820fd18b.gif)

**Comment by ljw1004** [comment]
@freebroccolo you wrote "there may be quite a lot of requests made in rapid succession and some of them may take some time to finish (e.g., find references). But in the meantime, other requests should be able to complete as soon as possible."

Are you sure about that? The LSP spec says [[link](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#request-notification-and-response-ordering)]

> Responses for requests should be sent in the same order as the requests appear on the server or client side. So for example if a server receives a textDocument/completion request and then a textDocument/signatureHelp request it should first return the response for the textDocument/completion and then the reponse for textDocument/signatureHelp.

LSP does lets you re-order notifications (so long as doing so doesn't affect correctness) but doesn't let you re-order requests+responses.

**Comment by freebroccolo** [comment]
@ljw1004 I am not talking about a completely arbitrary reordering of tasks in such a way that would produce incorrect behavior in the editor with regard to the protocol. From the text you quoted, the next paragraph states the following:

> How the server internally processes the requests is up to the server implementation. If the server decides to execute them in parallel and this produces correct result the server is free to do so. The server is also allowed to reorder requests and notification if the reordering doesn't affect correctness.

**Comment by ljw1004** [comment]
@freebroccolo Right, so it's allowed to re-order how it internally does the *processing*, but the *responses* to any request must still be sent back in the original order that the requests were received.

**Comment by freebroccolo** [comment]
I'm not particularly interested in debating in great detail the intent of those passages in that document because I think it's unclear as its currently written. I will just say that the way I read it is apparently not what you are suggesting. The way I have the OCaml language server currently implement is also not constrained in this fashion and it doesn't seem to be an issue.

**Comment by XVilka** [comment]
Was there any decision/update on this one? One year later LSP gained much more popularity than in 2017.

**Comment by andreypopp** [comment]
Should this be closed? As @bryphe's implementation is now upstreamed to merlin.

**Comment by rvantonder** [comment]
Oh, I just got caught up on these upstream changes. FYI @andreypopp, I see the [protocol code](https://github.com/ocaml/merlin/blob/master/src/lsp/protocol.ml) is taken from Flow. I (re)wrote a functor-based implementation after looking at the Flow code (using yojson from the beginning) that is currently used in [Pyre](https://github.com/facebook/pyre-check/tree/master/server/language_server). The reason for the rewrite was that the flow types actually don't correspond as closely to the spec as I'd like (e.g., 'experimental' types), and abstracting things like Messages/Requests/Responses to the module level works well. 

I actually pulled out that LSP code and created [its own package](https://github.com/rvantonder/ocaml-lsp-specification), example [here](https://github.com/rvantonder/ocaml-lsp-specification/blob/master/lib/types.ml). I've been meaning to finish it, but should have most of what's being used right now. IMO down the line it would be nice to have a library with the spec/types for reuse. Feel free to open a separate issue if you think this is working looking into.

**Comment by andreypopp** [comment]
@rvantonder yes, some came from @bryphe and then some from facebook/flow. I've also noticed that facebook/flow types are a bit off in some places.

I agree that we should have a common package for an LSP protocol spec! I'll take a look at rvantonder/ocaml-lsp-specification, it's already looking strictly better than what we have here as it doesn't use ppx deriving.

**Comment by Khady** [comment]
It does use ppx deriving. Or there is something important I missed.

**Comment by andreypopp** [comment]
@Khady indeed, I was wrong
